pallets/foreign-investments/src/entities.rs

```diff
@@ -98,131 +158,221 @@ impl<T: Config> InvestmentInfo<T> {
 	) -> Result<SwapOf<T>, DispatchError> {
 		let pool_currency = pool_currency_of::<T>(investment_id)?;
 
-		// NOTE: This line will be removed with market ratios
-		let pool_amount = T::CurrencyConverter::stable_to_stable(
+		// We do not want to decrease the whole `foreign_amount` from the investment
+		// amount if there is a pending investment swap.
+		let increasing_foreign_amount = self.pending_increase_swap(who, investment_id)?;
+		let foreign_investment_decrement = foreign_amount.saturating_sub(increasing_foreign_amount);
+
+		let pool_investment_decrement = self
+			.correlation
+			.foreign_to_pool(foreign_investment_decrement)
+			.map_err(|e| match e {
+				DispatchError::Arithmetic(ArithmeticError::DivisionByZero) => {
+					Error::<T>::TooMuchDecrease.into()
+				}
+				e => e,
+			})?;
+
+		self.decrease_investment(who, investment_id, pool_investment_decrement)?;
+
+		// It's ok to use the market ratio because this amount will be
+		// cancelled in this instant.
+		let increasing_pool_amount = T::TokenSwaps::convert_by_market(
 			pool_currency,
-			self.base.foreign_currency,
-			foreign_amount,
+			self.foreign_currency,
+			min(foreign_amount, increasing_foreign_amount),
 		)?;
 
-		let pending_pool_amount_increment =
-			Swaps::<T>::pending_amount_for(who, investment_id, Action::Investment, pool_currency);
-
-		let investment_decrement = pool_amount.saturating_sub(pending_pool_amount_increment);
-		if !investment_decrement.is_zero() {
-			T::Investment::update_investment(
-				who,
-				investment_id,
-				T::Investment::investment(who, investment_id)?
-					.ensure_sub(investment_decrement)
-					.map_err(|_| Error::<T>::TooMuchDecrease)?,
-			)?;
-		}
-
 		Ok(Swap {
-			currency_in: self.base.foreign_currency,
+			currency_in: self.foreign_currency,
 			currency_out: pool_currency,
-			amount_in: foreign_amount,
+			amount_out: increasing_pool_amount.ensure_add(pool_investment_decrement)?,
```

Comment:
I am looking at this code now for 4 hours. I thought multiple times I found a bug, but I always recovered later down the road.

I am now pretty certain I found a bug, as the `T::update_investment()` logic works different than the code expects here. It is updating the investment to the given value, so it expects an absolute value not a delta. So the bug is ONLY visible in integration tests.

### Example 1
1. `Increase(100 ethUSDC)`
1.1. `Swap(0 USDC, 100 ethUSDC)`
2. `FillOrder(70 USDC)`
2.1. `Swap(70 USDC, 30 ethUSDC)`
2.2. `Investment(70 USDC)`
3. `Decrease(20 ethUSDC)` 

#### `pre_decrease_swap(..)` 
* `foreign_amount` &rarr; `20`
* `increasing_foreign_amount  = self.pending_increase_swap(..)` &rarr;  `30`
* `foreign_investment_decrement = sat(foreign_amount - increasing_foreign_amount)` &rarr;  `sat(20-30) = 0`
* `self.decrease_investment(foreign_investment_decrement)` &rarr;  `Investment()` ❌  &rarr;  should be `Investment(70)`
* `increasing_pool_amount = min(foreign_amount, increasing_foreign_amount)` &rarr; `min(20, 30) = 20`
* `amount_out: increasing_pool_amount + foreign_investment_decrement` &rarr; `20 + 0 = 20` 
   
**Returned `SwapOf`**
* `amount_out: 20`  ✅   &rarr; swap will be updated as inverse with `Ordering::Smaller` &rarr;  `amount_out = 10`
* `currency_in: ethUSDC` ✅  
* `currency_out: USDC` ✅  
      
### Example 2 
1. `Increase(100 ethUSDC)`
1.1. `Swap(0 USDC, 100 ethUSDC)`
2. `FillOrder(70 USDC)`
2.1. `Swap(70 USDC, 30 ethUSDC)`
2.2. `Investment(70 USDC)`
3. `Decrease(100 ethUSDC)` 

#### `pre_decrease_swap(..)` 
* `foreign_amount` &rarr; `100`
* `increasing_foreign_amount  = self.pending_increase_swap(..)` &rarr;  `30`
* `foreign_investment_decrement = sat(foreign_amount - increasing_foreign_amount)` &rarr;  `sat(100-30) = 70`
* `self.decrease_investment(foreign_investment_decrement)` &rarr;  `Investment(70)` ❌  &rarr;  should be `Investment(0)` 
* `increasing_pool_amount = min(foreign_amount, increasing_foreign_amount)` &rarr; `min(100, 30) = 30`
* `amount_out: increasing_pool_amount + foreign_investment_decrement` &rarr; `30 + 70 = 100` 
   
**Returned `SwapOf`**
* `amount_out: 100` ✅  &rarr; swap will be updated as inverse with `Ordering::Greater` &rarr; `amount_out = 70`
* `currency_in: ethUSDC` ✅ 
* `currency_out: USDC` ✅ 

### Example 3
1. `Increase(100 ethUSDC)`
1.1. `Swap(0 USDC, 100 ethUSDC)`
2. `FillOrder(70 USDC)`
2.1. `Swap(70 USDC, 30 ethUSDC)`
2.2. `Investment(70 USDC)`
3. `Decrease(40 ethUSDC)` 

#### `pre_decrease_swap(..)` 
* `foreign_amount` &rarr; `40`
* `increasing_foreign_amount  = self.pending_increase_swap(..)` &rarr;  `30`
* `foreign_investment_decrement = sat(foreign_amount - increasing_foreign_amount)` &rarr;  `sat(40-30) = 10`
* `self.decrease_investment(foreign_investment_decrement)` &rarr;  `Investment(10)` ❌  &rarr;  should be `Investment(20)`
* `increasing_pool_amount = min(foreign_amount, increasing_foreign_amount)` &rarr; `min(40, 30) = 30`
* `amount_out: increasing_pool_amount + foreign_investment_decrement` &rarr; `30 + 10= 50` 
   
**Returned `SwapOf`**
* `amount_out: 50` ✅  &rarr; swap will be updated as inverse with `Ordering::Greater` &rarr; `amount_out = 20`
* `currency_in: ethUSDC` ✅ 
* `currency_out: USDC` ✅

Reply:
Ah fuck! Haha. Yeah, found that line! That its fine.

Reply:
And nevermind regarding the complexity.